### PR TITLE
Fermichem part 2.1: Adds a reaction/reagent lookup GUI!

### PR DESCRIPTION
**Pull request by Thalpy**
## About The Pull Request

Adds a new GUI that can be opened from the chem dispenser that will let you look up all reactions and reagents presently in the game (except secret recipes).

For the colours:
purple - clicking this will give you a recipe
blue - clicking this will give you a reagent
green - clicking this will turn it off (or it's also green if the reagent is present in the associated beaker the UI is linked to)
red - clicking this will turn it on 

I tend to get verbose so here's a video instead!

https://user-images.githubusercontent.com/33956696/107724825-aca48480-6cdc-11eb-9ba3-800627fc937e.mp4

And a picture:
image

## Why It's Good For The Game

Fermichem changed a lot, and I want to make it possible for people to see reaction conditions easily rather than having to alt tab to a wiki constantly. I don't think this will replace the wiki though - this is more a condensed version for finding reaction conditions, whereas the wiki has more information.

## Changelog
:cl:
add: Adds a new reaction lookup window to the chem dispenser
/:cl:

### CODE REVIEW

**Comment by Thalpy** [comment]
I'm having a few troubles with .js linters - I've 1 too many consts, and one of my lines is 1 char over. I figure I'll likely have to change bits about it anyways because it's a define dump to deal with bitflags, so hopefully they'll be cleaned up from feedback.

```
     7:25  error  Arrow function has a complexity of 51. Maximum allowed is 50  complexity
  522:1   error  This line has a length of 81. Maximum allowed is 80           max-len
```

**Comment by Hamcha** [comment]
As someone who spent a lot of time making a better wiki client just because of how vital having a reference is, I cannot thank you enough for this!

However, I feel like for more complex reagents it might be just as finnicky to lookup parent reagents as vanilla wiki is.
In my script, I auto-nested all recipes like this:



It is not pretty but in my experience it's way too handy.

**Comment by Thalpy** [comment]
> 
> 
> As someone who spent a lot of time making a better wiki client just because of how vital having a reference is, I cannot thank you enough for this!
> 
> However, I feel like for more complex recipes it might be just as finnicky to lookup parent reagents as vanilla wiki is.
> In my script, I auto-nested all recipes like this:
> 
> 
> 
> It is not pretty but in my experience it's way too handy.

I wanted to do something similar to that, but I couldn't figure out how to have miltiple tooltips appear to the right of a tooltip (i.e. nesting tooltips). Do you have any idea how to do that? Presently it shows one level down as a tooltip hover.

**Comment by Hamcha** [comment]
Multiple tooltips not working is why I moved away from tooltips. (In fact, most of the work I did on the script was just making the tooltips *not* tooltips and inline them to the content of the page).

I have near zero experience with tgui, so no idea how to help you there, but from a design perspective: perhaps some sort of bookmark system might help alleviate the issue?

File: tgui/packages/tgui/interfaces/Reagents.js
Comment (Jared-Fogle) [inline tgui/packages/tgui/interfaces/Reagents.js:27]: These should be passed down through `ui_static_data` so we don't have two places to update if we want to change the flags.

Comment (stylemistake) [reply tgui/packages/tgui/interfaces/Reagents.js:27]: true

Comment (Thalpy) [reply tgui/packages/tgui/interfaces/Reagents.js:27]: Hmm, do I still set them up as consts then?

**Comment by Thalpy** [comment]
> 
> 
> Multiple tooltips not working is why I moved away from tooltips. (In fact, most of the work I did on the script was just making the tooltips _not_ tooltips and inline them to the content of the page).
> 
> I have near zero experience with tgui, so no idea how to help you there, but from a design perspective: perhaps some sort of bookmark system might help alleviate the issue? If you could jump back to your main recipe while seeing the nested reactions, not having a tree view probably becomes a non-issue (and it would work better with beaker sync imo)

I like the idea of bookmarks, I could just get a big list of all component reagents, but with how reactions work now it would likely not work out if you blindly followed it (since everything will react if it gets the chance). I'll try to put that in when I've reworked the js code from maintainer feedback.

File: tgui/packages/tgui/interfaces/Reagents.js
Comment (stylemistake) [reply tgui/packages/tgui/interfaces/Reagents.js:27]: return them via ui_static_data

**Comment by Tlaltecuhtli** [comment]
can the recipe you want to do be like uhh shown on the window with main chem dispenser window once clicked? like you click ephedrine and the main window shows 1 oil 1 dieth 1 sugar 1 hydrogen as suggestion which you can click X to remove it once ur done

**Comment by Thalpy** [comment]
Okay, I added a bookmarks function. It saves itself between instances and beakers too!

https://user-images.githubusercontent.com/33956696/107855816-88e85800-6e1c-11eb-8705-29c00df996e4.mp4

> 
> 
> can the recipe you want to do be like uhh shown on the window with main chem dispenser window once clicked? like you click ephedrine and the main window shows 1 oil 1 dieth 1 sugar 1 hydrogen as suggestion which you can click X to remove it once ur done

Hmm, I'm not sure how to fit it into the dispenser ui cleanly, but I have a similar idea, I'll just get maintainer approval first.

**Comment by Thalpy** [comment]
Okay so instead of having recipes appear on the chem dispenser, I've instead made it so that if the dispenser has a reagent in the selected recipe it will highlight it.

File: tgui/packages/tgui/interfaces/common/ReagentLookup.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/ReagentLookup.js:121]: Use early return in this component.

```
if (!reagents) {
  return (
    your box
  );
}

return (
  rest of UI
);
```

File: tgui/packages/tgui/interfaces/common/ReagentLookup.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/ReagentLookup.js:22]:
```suggestion
            onClick={() => {
              Byond.command(`wiki Guide_to_chemistry#${reagent.name}`);
            }} />
```

File: tgui/packages/tgui/interfaces/common/ReagentLookup.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/ReagentLookup.js:65]:
```suggestion
                  content={reagent.inverseReagent}
```

File: tgui/packages/tgui/interfaces/common/RecipeLookup.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/RecipeLookup.js:207]: Same can be done here, good candidate for early return.

File: tgui/packages/tgui/interfaces/common/RecipeLookup.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/RecipeLookup.js:173]: Break up into multiple lines. Linter ignores these lines because they have strings, but it doesn't exempt you from at least putting newlines before `?` and `:`.

File: tgui/packages/tgui/interfaces/common/RecipeLookup.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/RecipeLookup.js:181]: Same here

File: tgui/packages/tgui/interfaces/common/RecipeLookup.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/RecipeLookup.js:28]: Are you serious?

`? true : false` moment.

```suggestion
            disabled={recipe.subReactIndex === recipe.subReactLen}
```

File: tgui/packages/tgui/interfaces/common/RecipeLookup.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/RecipeLookup.js:19]:
```suggestion
            disabled={recipe.subReactIndex === 1}
```

File: tgui/packages/tgui/interfaces/common/RecipeLookup.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/common/RecipeLookup.js:15]: Pointless key
```suggestion
```

File: tgui/packages/tgui/interfaces/ChemDispenser.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/ChemDispenser.js:149]: break up this long line on `?` and `:`

**Review by stylemistake — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/common/ReagentLookup.js
Comment (Thalpy) [reply tgui/packages/tgui/interfaces/common/ReagentLookup.js]: Done so

File: tgui/packages/tgui/interfaces/common/RecipeLookup.js
Comment (Thalpy) [reply tgui/packages/tgui/interfaces/common/RecipeLookup.js]: same as above

File: tgui/packages/tgui/interfaces/common/RecipeLookup.js
Comment (Thalpy) [reply tgui/packages/tgui/interfaces/common/RecipeLookup.js]: Didn't realise I could do that! Thanks.

File: tgui/packages/tgui/interfaces/common/RecipeLookup.js
Comment (Thalpy) [reply tgui/packages/tgui/interfaces/common/RecipeLookup.js]: same as above

File: tgui/packages/tgui/interfaces/ChemDispenser.js
Comment (Thalpy) [reply tgui/packages/tgui/interfaces/ChemDispenser.js]: Done

File: code/__DEFINES/reagents.dm
Comment (Mothblocks) [inline code/__DEFINES/reagents.dm:20]: s

Comment (Thalpy) [reply code/__DEFINES/reagents.dm]: ctrl s without the ctrl

File: tgui/packages/tgui/interfaces/ChemDispenser.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/ChemDispenser.js:39]: Replace with an empty tag, <>, </>

File: tgui/packages/tgui/interfaces/Reagents.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Reagents.js:20]: you can replace all of this function with currentReagents?.includes(reagent)

File: tgui/packages/tgui/interfaces/Reagents.js
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Reagents.js:79]: Spaces

File: tgui/packages/tgui/interfaces/ChemDispenser.js
Comment (Thalpy) [reply tgui/packages/tgui/interfaces/ChemDispenser.js]: Done

File: tgui/packages/tgui/interfaces/Reagents.js
Comment (Thalpy) [reply tgui/packages/tgui/interfaces/Reagents.js]: done

File: tgui/packages/tgui/interfaces/Reagents.js
Comment (Thalpy) [reply tgui/packages/tgui/interfaces/Reagents.js]: And done!

**Comment by stylemistake** [comment]
Hey, master of copypasta, please check if this still works as intended.

Review my changes and maybe you'll learn something.

What I don't like about your code in general:
- You use Button as a glorified holder for a tooltip that otherwise does nothing. You override a lot of props just to make it look like NOT a button. There is a way to use a tooltip without a button.
- You do not organize imports, there is a menu item for that in VSCode (right click, source action, organize imports)
- Using keys where there is no need for keys (e.g. not in a loop)
- Not using keys where there IS a need for keys (e.g. in a loop, this was the reason why after I edited some of your code, linter has actually started complaining about missing keys)
- Using `"white"`, when theme default font color is already white. You ideally should not be using `"white"`

How you do it with white:
```
color={bleh ? 'red' : 'white'}
```

How you should be doing it:
```
color={bleh && 'red'}
```

**Review by stylemistake — APPROVED** [review]

**Comment by drongood12** [comment]
The window with the reagents opens for a very long time due to the fact that all reagents are being displayed, you probably should have also added pagination, and the search by partial match is also missing. I would also like to see the close button in the Recipe lookup and Reagent lookup menu, because sometimes they interfere with the selection of the desired reagent or recipe

**Comment by Thalpy** [comment]
> 
> 
> The window with the reagents opens for a very long time due to the fact that all reagents are being displayed, you probably should have also added pagination, and the search by partial match is also missing. I would also like to see the close button in the Recipe lookup and Reagent lookup menu, because sometimes they interfere with the selection of the desired reagent or recipe
> image

I'll see if I can figure out pagination, and add a close button while I'm at it. I'm not sure what you mean by partial match though.

**Comment by drongood12** [comment]
> > The window with the reagents opens for a very long time due to the fact that all reagents are being displayed, you probably should have also added pagination, and the search by partial match is also missing. I would also like to see the close button in the Recipe lookup and Reagent lookup menu, because sometimes they interfere with the selection of the desired reagent or recipe
> > image
> 
> I'll see if I can figure out pagination, and add a close button while I'm at it. I'm not sure what you mean by partial match though. The UI was less laggy when I was using it locally unforunately.

By partial coincidence, I meant that you could enter, for example, space and I was given all the chemicals that contain this text. Also why does a chemical dispenser need bartender drinks? Why not make a separation for this, the chemist only has reagents, the barmern only has drinks

**Comment by Thalpy** [comment]
> 
> 
> > > The window with the reagents opens for a very long time due to the fact that all reagents are being displayed, you probably should have also added pagination, and the search by partial match is also missing. I would also like to see the close button in the Recipe lookup and Reagent lookup menu, because sometimes they interfere with the selection of the desired reagent or recipe
> > > image
> > 
> > 
> > I'll see if I can figure out pagination, and add a close button while I'm at it. I'm not sure what you mean by partial match though. The UI was less laggy when I was using it locally unforunately.
> 
> By partial match, I mean that you can enter, for example, space and all the chemicals in the list containing that text will be displayed. Also why does a chemical dispenser need a bartender's drinks? Why not make a separation for this, the chemist has only reagents, the bartender has only drinks
> 
> P.s
> 
> I may be mistaken in writing some sentences, because English is not my native language. Sorry(

Don't worry! Your english is understandable and I get what you mean since you clarified. To do the search as a list thing might take a bit more time, and I want to make sure some of the more important parts of chemistry are added before I attempt that, but I did add the pagination and closing request you put in. I figured I might as well have all recipies in one place for ease of use and incase bartenders/chemists felt like broadening their horizons since it's meant to be a lookup thing.

**Pull request merged**
